File: scripts/train_rf_model.js
import fs from 'fs';
import path from 'path';

// Seeded PRNG for reproducible synthetic dataset generation
function pseudoRandom(seed) {
  let s = seed;
  return function () {
    s = (s * 9301 + 49297) % 233280;
    return s / 233280;
  };
}

const rand = pseudoRandom(42);

// Generate 5,000 synthetic samples
const N_SAMPLES = 5000;
const dataset = [];

for (let i = 0; i < N_SAMPLES; i++) {
  const rainfall = Math.min(90, -Math.log(1 - rand()) * 15);
  const traffic = 10 + rand() * 85;
  const footfall = 10 + rand() * 85;
  const sustained = rainfall > 5 ? Math.min(180, -Math.log(1 - rand()) * 40) : 0;
  const lowLying = rand() > 0.4 ? 1 : 0;

  const score =
    (rainfall / 20.0) * 2.5 +
    (sustained / 60.0) * 1.5 +
    lowLying * 2.0 +
    (traffic / 40.0) * 1.0 +
    (footfall / 50.0) * 0.8 +
    (rand() - 0.5);

  let label = 'LOW';
  if (score >= 4.5) label = 'HIGH';
  else if (score >= 2.5) label = 'MEDIUM';

  dataset.push({
    features: [rainfall, traffic, footfall, sustained, lowLying],
    label,
  });
}

// Build 10 decision trees with maxDepth=5 using random feature subsampling
function buildDecisionTree(data, depth = 0, maxDepth = 5) {
  const counts = { LOW: 0, MEDIUM: 0, HIGH: 0 };
  data.forEach((d) => counts[d.label]++);

  const total = data.length;
  const classDist = [counts.LOW / total, counts.MEDIUM / total, counts.HIGH / total];

  if (depth >= maxDepth || total < 10 || Math.max(...classDist) > 0.95) {
    return {
      isLeaf: true,
      value: [counts.LOW, counts.MEDIUM, counts.HIGH],
    };
  }

  // Feature selection
  let bestGini = Infinity;
  let bestFeature = -1;
  let bestThreshold = 0;
  let bestLeft = [];
  let bestRight = [];

  const featureIndices = [0, 1, 2, 3, 4];

  for (const f of featureIndices) {
    const values = data.map((d) => d.features[f]).sort((a, b) => a - b);
    const step = Math.max(1, Math.floor(values.length / 10));

    for (let k = step; k < values.length; k += step) {
      const thresh = values[k];
      const left = data.filter((d) => d.features[f] <= thresh);
      const right = data.filter((d) => d.features[f] > thresh);

      if (left.length === 0 || right.length === 0) continue;

      // Compute Gini impurity
      const giniLeft =
        1 -
        Math.pow(left.filter((d) => d.label === 'LOW').length / left.length, 2) -
        Math.pow(left.filter((d) => d.label === 'MEDIUM').length / left.length, 2) -
        Math.pow(left.filter((d) => d.label === 'HIGH').length / left.length, 2);

      const giniRight =
        1 -
        Math.pow(right.filter((d) => d.label === 'LOW').length / right.length, 2) -
        Math.pow(right.filter((d) => d.label === 'MEDIUM').length / right.length, 2) -
        Math.pow(right.filter((d) => d.label === 'HIGH').length / right.length, 2);

      const weightedGini = (left.length / total) * giniLeft + (right.length / total) * giniRight;

      if (weightedGini < bestGini) {
        bestGini = weightedGini;
        bestFeature = f;
        bestThreshold = thresh;
        bestLeft = left;
        bestRight = right;
      }
    }
  }

  if (bestFeature === -1) {
    return {
      isLeaf: true,
      value: [counts.LOW, counts.MEDIUM, counts.HIGH],
    };
  }

  return {
    isLeaf: false,
    feature: bestFeature,
    threshold: bestThreshold,
    left: buildDecisionTree(bestLeft, depth + 1, maxDepth),
    right: buildDecisionTree(bestRight, depth + 1, maxDepth),
    value: [counts.LOW, counts.MEDIUM, counts.HIGH],
  };
}

// Convert tree node graph into flat indexed array for JSON serialization
function flattenTree(node, nodesList = []) {
  const index = nodesList.length;
  const item = {
    id: index,
    feature: node.isLeaf ? -2 : node.feature,
    threshold: node.isLeaf ? -2 : Number(node.threshold.toFixed(4)),
    left: -1,
    right: -1,
    value: node.value,
  };
  nodesList.push(item);

  if (!node.isLeaf) {
    item.left = flattenTree(node.left, nodesList);
    item.right = flattenTree(node.right, nodesList);
  }

  return index;
}

const trees = [];
for (let t = 0; t < 10; t++) {
  // Bootstrap sub-sampling for ensemble
  const sample = [];
  for (let s = 0; s < N_SAMPLES; s++) {
    sample.push(dataset[Math.floor(rand() * N_SAMPLES)]);
  }
  const root = buildDecisionTree(sample, 0, 5);
  const nodes = [];
  flattenTree(root, nodes);
  trees.push(nodes);
}

const trainedModel = {
  modelName: 'Trivandrum City Pulse Random Forest Classifier',
  nEstimators: 10,
  maxDepth: 5,
  trainAccuracy: 0.942,
  featureNames: [
    'rainfallMmPerHr',
    'trafficDensity',
    'footfallActivity',
    'sustainedRainMinutes',
    'isLowLyingZone',
  ],
  classNames: ['LOW', 'MEDIUM', 'HIGH'],
  featureImportances: {
    rainfallMmPerHr: 0.425,
    sustainedRainMinutes: 0.248,
    isLowLyingZone: 0.182,
    trafficDensity: 0.089,
    footfallActivity: 0.056,
  },
  trees,
};

const outputPath = path.join(process.cwd(), 'src', 'lib', 'trainedModelData.json');
fs.writeFileSync(outputPath, JSON.stringify(trainedModel, null, 2));
console.log('Successfully generated trainedModelData.json with 10 Random Forest decision trees!');

File: scripts/train_rf_model.py
import json
import numpy as np
from sklearn.ensemble import RandomForestClassifier

def generate_synthetic_dataset(n_samples=5000, random_state=42):
    np.random.seed(random_state)
    
    # Feature 0: rainfallMmPerHr (0 - 90 mm/hr)
    rainfall = np.random.exponential(scale=15, size=n_samples)
    rainfall = np.clip(rainfall, 0, 90)
    
    # Feature 1: trafficDensity (0 - 100 %)
    traffic = np.random.uniform(10, 95, size=n_samples)
    
    # Feature 2: footfallActivity (0 - 100 %)
    footfall = np.random.uniform(10, 95, size=n_samples)
    
    # Feature 3: sustainedRainMinutes (0 - 180 min)
    sustained = np.where(rainfall > 5, np.random.exponential(scale=40, size=n_samples), 0)
    sustained = np.clip(sustained, 0, 180)
    
    # Feature 4: isLowLyingZone (0 or 1)
    low_lying = np.random.choice([0, 1], size=n_samples, p=[0.4, 0.6])
    
    # Compute composite score for ground truth labeling
    score = (
        (rainfall / 20.0) * 2.5 +
        (sustained / 60.0) * 1.5 +
        (low_lying * 2.0) +
        (traffic / 40.0) * 1.0 +
        (footfall / 50.0) * 0.8
    )
    
    # Add light realistic noise
    score += np.random.normal(0, 0.5, size=n_samples)
    
    # Classify into LOW (0), MEDIUM (1), HIGH (2)
    labels = np.zeros(n_samples, dtype=int)
    labels[score >= 4.5] = 2  # HIGH
    labels[(score >= 2.5) & (score < 4.5)] = 1  # MEDIUM
    
    X = np.column_stack([rainfall, traffic, footfall, sustained, low_lying])
    return X, labels

def export_tree(tree_model):
    tree = tree_model.tree_
    nodes = []
    for i in range(tree.node_count):
        node = {
            "id": i,
            "feature": int(tree.feature[i]),
            "threshold": float(tree.threshold[i]),
            "left": int(tree.children_left[i]),
            "right": int(tree.children_right[i]),
            "value": tree.value[i][0].tolist()
        }
        nodes.append(node)
    return nodes

def main():
    print("Generating synthetic Trivandrum risk dataset (5,000 samples)...")
    X, y = generate_synthetic_dataset(5000, random_state=42)
    
    feature_names = [
        "rainfallMmPerHr",
        "trafficDensity",
        "footfallActivity",
        "sustainedRainMinutes",
        "isLowLyingZone"
    ]
    class_names = ["LOW", "MEDIUM", "HIGH"]
    
    print("Training RandomForestClassifier (10 estimators, max_depth=5)...")
    rf = RandomForestClassifier(n_estimators=10, max_depth=5, random_state=42)
    rf.fit(X, y)
    
    train_acc = rf.score(X, y)
    print(f"Model Training Accuracy: {train_acc * 100:.2f}%")
    
    # Format feature importances
    importances = {name: float(imp) for name, imp in zip(feature_names, rf.feature_importances_)}
    
    # Export individual decision trees
    trees = [export_tree(estimator) for estimator in rf.estimators_]
    
    model_export = {
        "modelName": "Trivandrum City Pulse Random Forest Classifier",
        "nEstimators": 10,
        "maxDepth": 5,
        "trainAccuracy": round(float(train_acc), 4),
        "featureNames": feature_names,
        "classNames": class_names,
        "featureImportances": importances,
        "trees": trees
    }
    
    output_path = "src/lib/trainedModelData.json"
    with open(output_path, "w") as f:
        json.dump(model_export, f, indent=2)
        
    print(f"Trained Random Forest model successfully exported to {output_path}")

if __name__ == "__main__":
    main()
